### PR TITLE
Remove background processing library dependency.

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/123764

Tickets:
* https://central.tri.be/issues/123764
* https://central.tri.be/issues/120795

This PR:
* removes the dependency on the `a5hleyrich/wp-background-processing` library
* internalizes the base, non overridden methods
* fixes an issue caused by a race condition caused, in turn, by the use of the sitemeta table
* fixes an issue where the `complete` method would not be called on empty queues
* fixes an issue where the `complete` method would not be called correctly during cron-based processing.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Queue.php:758](https://github.com/moderntribe/tribe-common/blob/eb911662360c032d2371d5d576430040cc2d86b7/src/Tribe/Process/Queue.php#L758)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Handler.php:161](https://github.com/moderntribe/tribe-common/blob/3da93bf334b395a2be0d8f3485317e0edca8a8ad/src/Tribe/Process/Handler.php#L161)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Handler.php:304](https://github.com/moderntribe/tribe-common/blob/3da93bf334b395a2be0d8f3485317e0edca8a8ad/src/Tribe/Process/Handler.php#L304)
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/3da93bf334b395a2be0d8f3485317e0edca8a8ad/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/3da93bf334b395a2be0d8f3485317e0edca8a8ad/src/Tribe/Process/Queue.php#L703)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Queue.php:757](https://github.com/moderntribe/tribe-common/blob/3da93bf334b395a2be0d8f3485317e0edca8a8ad/src/Tribe/Process/Queue.php#L757)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Handler.php:161](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Handler.php#L161)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Handler.php:304](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Handler.php#L304)
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Queue.php#L703)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Queue.php:757](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Queue.php#L757)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/f3d6a84bdb3fbe5f0da94834e1853787ad6f9cc4/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Handler.php:161](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Handler.php#L161)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Handler.php:304](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Handler.php#L304)
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Queue.php#L703)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Queue.php:757](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Queue.php#L757)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Queue.php:1094](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Queue.php#L1094)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/a357f5216e901cd9e63830bb6ce00e0d43256306/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Handler.php:161](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Handler.php#L161)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Handler.php:304](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Handler.php#L304)
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Queue.php#L703)
:x: | Scope keyword &quot;protected&quot; must be followed by a single space | [Queue.php:757](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Queue.php#L757)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Queue.php:1094](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Queue.php#L1094)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/48ca9df014ae2f25695bb188dec093dbc3cdc2a3/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/5dce68de190f1dbc78bce00733544241962e8dbd/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/5dce68de190f1dbc78bce00733544241962e8dbd/src/Tribe/Process/Queue.php#L703)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/5dce68de190f1dbc78bce00733544241962e8dbd/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Queue.php:1094](https://github.com/moderntribe/tribe-common/blob/5dce68de190f1dbc78bce00733544241962e8dbd/src/Tribe/Process/Queue.php#L1094)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/5dce68de190f1dbc78bce00733544241962e8dbd/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | A cast statement must be followed by a single space | [Queue.php:545](https://github.com/moderntribe/tribe-common/blob/8319fbba2173abc49768c819ea0d9a79b335f7a9/src/Tribe/Process/Queue.php#L545)
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/8319fbba2173abc49768c819ea0d9a79b335f7a9/src/Tribe/Process/Queue.php#L703)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/8319fbba2173abc49768c819ea0d9a79b335f7a9/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Queue.php:1094](https://github.com/moderntribe/tribe-common/blob/8319fbba2173abc49768c819ea0d9a79b335f7a9/src/Tribe/Process/Queue.php#L1094)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/8319fbba2173abc49768c819ea0d9a79b335f7a9/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:703](https://github.com/moderntribe/tribe-common/blob/8ce3db4600ffdbda4cee3ea149f724b1cfe3df58/src/Tribe/Process/Queue.php#L703)
:warning: | The method parameter $data_source is never used | [Queue.php:856](https://github.com/moderntribe/tribe-common/blob/8ce3db4600ffdbda4cee3ea149f724b1cfe3df58/src/Tribe/Process/Queue.php#L856)
:warning: | The method parameter $data_source is never used | [Queue.php:1094](https://github.com/moderntribe/tribe-common/blob/8ce3db4600ffdbda4cee3ea149f724b1cfe3df58/src/Tribe/Process/Queue.php#L1094)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/8ce3db4600ffdbda4cee3ea149f724b1cfe3df58/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:675](https://github.com/moderntribe/tribe-common/blob/e9aeee577aed7c3e80c70bde5ce7d7419cbdbefe/src/Tribe/Process/Queue.php#L675)
:warning: | The method parameter $data_source is never used | [Queue.php:820](https://github.com/moderntribe/tribe-common/blob/e9aeee577aed7c3e80c70bde5ce7d7419cbdbefe/src/Tribe/Process/Queue.php#L820)
:warning: | The method parameter $data_source is never used | [Queue.php:1058](https://github.com/moderntribe/tribe-common/blob/e9aeee577aed7c3e80c70bde5ce7d7419cbdbefe/src/Tribe/Process/Queue.php#L1058)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/e9aeee577aed7c3e80c70bde5ce7d7419cbdbefe/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:682](https://github.com/moderntribe/tribe-common/blob/f187aa3b0b117cb77c00a31e148f69096211d189/src/Tribe/Process/Queue.php#L682)
:warning: | The method parameter $data_source is never used | [Queue.php:830](https://github.com/moderntribe/tribe-common/blob/f187aa3b0b117cb77c00a31e148f69096211d189/src/Tribe/Process/Queue.php#L830)
:warning: | The method parameter $data_source is never used | [Queue.php:1077](https://github.com/moderntribe/tribe-common/blob/f187aa3b0b117cb77c00a31e148f69096211d189/src/Tribe/Process/Queue.php#L1077)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/f187aa3b0b117cb77c00a31e148f69096211d189/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Process/Handler.php [hunk 1]
```diff
@@ -6,71 +6,95 @@
...
-		$this->action = call_user_func( array( $class, 'action' ) );
-		parent::__construct();
+		$class            = get_class( $this );
+		$this->action     = call_user_func( [ $class, 'action' ] );
+		$this->identifier = $this->prefix . '_' . $this->action;
 
-		$this->cron_hook_identifier = $this->identifier;
-		$this->feature_detection = tribe( 'feature-detection' );
+		add_action( 'wp_ajax_' . $this->identifier, [ $this, 'maybe_handle' ] );
+		add_action( 'wp_ajax_nopriv_' . $this->identifier, [ $this, 'maybe_handle' ] );
```
Comment (bordoni) [inline src/Tribe/Process/Handler.php]: We should add a possible way of users deactivating that for customers.

File: src/Tribe/Process/Handler.php [hunk 2]
```diff
@@ -85,63 +109,155 @@ public function __construct() {
...
+	 * @since TBD Pulled from the `WP_Async_Request` class.
+	 *
+	 * @return array An array of arguments for the POST request.
+	 */
+	protected function get_query_args() {
+		if ( null !== $this->query_args ) {
+			return $this->query_args;
+		}
+
+		return array(
```
Comment (bordoni) [inline src/Tribe/Process/Handler.php]: Use `[]` notation

File: src/Tribe/Process/Handler.php [hunk 3]
```diff
@@ -85,63 +109,155 @@ public function __construct() {
...
+	 * @since TBD Pulled from the `WP_Async_Request` class.
+	 *
+	 * @return array An array of arguments that will be used to send the POST request.
+	 */
+	protected function get_post_args() {
+		if ( null !== $this->post_args ) {
+			return $this->post_args;
+		}
+
+		return array(
```
Comment (bordoni) [inline src/Tribe/Process/Handler.php]: Use `[]` notation

File: src/Tribe/Process/Queue.php [hunk 4]
```diff
@@ -71,6 +124,12 @@ public function __construct() {
 
 		parent::__construct();
 
+		$this->healthcheck_cron_hook_id     = $this->identifier . '_cron';
+		$this->healthcheck_cron_interval_id = $this->identifier . '_cron_interval';
+
+		add_action( $this->healthcheck_cron_hook_id, array( $this, 'handle_cron_healthcheck' ) );
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Use `[]` notation

File: src/Tribe/Process/Queue.php [hunk 5]
```diff
@@ -71,6 +124,12 @@ public function __construct() {
...
+		add_filter( 'cron_schedules', array( $this, 'schedule_cron_healthcheck' ) );
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Use `[]` notation

File: src/Tribe/Process/Queue.php [hunk 6]
```diff
@@ -602,26 +671,411 @@ protected function get_post_args() {
...
+	protected function memory_exceeded() {
+		$memory_limit   = $this->get_memory_limit() * 0.9; // 90% of max memory
+		$current_memory = memory_get_usage( true );
+		$return         = false;
+
+		if ( $current_memory >= $memory_limit ) {
+			$return = true;
+		}
+
+		return apply_filters( $this->identifier . '_memory_exceeded', $return );
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Docblock for filter.

File: src/Tribe/Process/Queue.php [hunk 7]
```diff
@@ -602,26 +671,411 @@ protected function get_post_args() {
...
+	 *
+	 * Ensures the batch never exceeds a sensible time limit.
+	 * A timeout limit of 30s is common on shared hosting.
+	 *
+	 * @since TBD Pulled from the `WP_Background_Process` class.
+	 *
+	 * @return bool Whether the execution time was exceeded or not.
+	 */
+	protected function time_exceeded() {
+		$finish = $this->start_time + apply_filters( $this->identifier . '_default_time_limit', 20 ); // 20 seconds
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Docblock for filter

File: src/Tribe/Process/Queue.php [hunk 8]
```diff
@@ -602,26 +671,411 @@ protected function get_post_args() {
...
+		$return = false;
+
+		if ( time() >= $finish ) {
+			$return = true;
+		}
+
+		return apply_filters( $this->identifier . '_time_exceeded', $return );
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Docblock filter

File: src/Tribe/Process/Queue.php [hunk 9]
```diff
@@ -602,26 +671,411 @@ protected function get_post_args() {
...
+	 * Filters the `cron_schedules` filter to add a check every 5 minutes.
+	 *
+	 * @since TBD Pulled from the `WP_Background_Process` class.
+	 *
+	 * @param mixed $schedules The cron schedules to check.
+	 *
+	 * @return mixed The updated cron schedules.
+	 */
+	public function schedule_cron_healthcheck( $schedules ) {
+		$interval = apply_filters( $this->identifier . '_cron_interval', $this->healthcheck_cron_interval );
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Docblock for filter

File: src/Tribe/Process/Queue.php [hunk 10]
```diff
@@ -602,26 +671,411 @@ protected function get_post_args() {
...
+
+		// Adds every 5 minutes to the existing schedules.
+		$schedules[ $this->identifier . '_cron_interval' ] = array(
+			'interval' => MINUTE_IN_SECONDS * $interval,
+			'display'  => sprintf( __( 'Every %d Minutes' ), $interval ),
```
Comment (bordoni) [inline src/Tribe/Process/Queue.php]: Add domain.

File: vendor/composer/autoload_commands_classmap.php [hunk 11]
```diff
@@ -6,4 +6,100 @@
 $baseDir = dirname($vendorDir);
 
 return array(
+    'CLI_Command' => $vendorDir . '/wp-cli/wp-cli/php/commands/src/CLI_Command.php',
```
Comment (bordoni) [inline vendor/composer/autoload_commands_classmap.php]: Should we be pushing this into customers?

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:682](https://github.com/moderntribe/tribe-common/blob/10699b3dea7e509d6ed281aaf3438eac0a3e236d/src/Tribe/Process/Queue.php#L682)
:warning: | The method parameter $data_source is never used | [Queue.php:830](https://github.com/moderntribe/tribe-common/blob/10699b3dea7e509d6ed281aaf3438eac0a3e236d/src/Tribe/Process/Queue.php#L830)
:warning: | The method parameter $data_source is never used | [Queue.php:1077](https://github.com/moderntribe/tribe-common/blob/10699b3dea7e509d6ed281aaf3438eac0a3e236d/src/Tribe/Process/Queue.php#L1077)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/10699b3dea7e509d6ed281aaf3438eac0a3e236d/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:682](https://github.com/moderntribe/tribe-common/blob/1024af6a439b600dadabc50d0cda9c570abb9328/src/Tribe/Process/Queue.php#L682)
:warning: | The method parameter $data_source is never used | [Queue.php:830](https://github.com/moderntribe/tribe-common/blob/1024af6a439b600dadabc50d0cda9c570abb9328/src/Tribe/Process/Queue.php#L830)
:warning: | The method parameter $data_source is never used | [Queue.php:1077](https://github.com/moderntribe/tribe-common/blob/1024af6a439b600dadabc50d0cda9c570abb9328/src/Tribe/Process/Queue.php#L1077)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/1024af6a439b600dadabc50d0cda9c570abb9328/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

File: vendor/composer/autoload_commands_classmap.php (hunk 11 above)
Comment (lucatume) [reply vendor/composer/autoload_commands_classmap.php]: No, that's been added due to a dev-including autoload dump generation, fixed.

**Comment by lucatume** [comment]
Ok, made a bunch of changes and this is ready for another look.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:682](https://github.com/moderntribe/tribe-common/blob/27349369fa2b8e457d515e223be7338fad1f9a04/src/Tribe/Process/Queue.php#L682)
:warning: | The method parameter $data_source is never used | [Queue.php:830](https://github.com/moderntribe/tribe-common/blob/27349369fa2b8e457d515e223be7338fad1f9a04/src/Tribe/Process/Queue.php#L830)
:warning: | The method parameter $data_source is never used | [Queue.php:1106](https://github.com/moderntribe/tribe-common/blob/27349369fa2b8e457d515e223be7338fad1f9a04/src/Tribe/Process/Queue.php#L1106)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/27349369fa2b8e457d515e223be7338fad1f9a04/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $data_source is never used | [Queue.php:682](https://github.com/moderntribe/tribe-common/blob/d235d5ba06ff1c46c93a18c1b821448dfbe4d84c/src/Tribe/Process/Queue.php#L682)
:warning: | The method parameter $data_source is never used | [Queue.php:841](https://github.com/moderntribe/tribe-common/blob/d235d5ba06ff1c46c93a18c1b821448dfbe4d84c/src/Tribe/Process/Queue.php#L841)
:warning: | The method parameter $data_source is never used | [Queue.php:1125](https://github.com/moderntribe/tribe-common/blob/d235d5ba06ff1c46c93a18c1b821448dfbe4d84c/src/Tribe/Process/Queue.php#L1125)
:warning: | The method parameter $data_source is never used | [Tester.php:83](https://github.com/moderntribe/tribe-common/blob/d235d5ba06ff1c46c93a18c1b821448dfbe4d84c/src/Tribe/Process/Tester.php#L83)

> via jenkins/codesniffer-phpcs

**Pull request merged**